File: INTeract-ive-Agent/typescript/src/framework/orchestration/supervisor.ts
/**
 * Supervisor Pattern for Multi-Agent Orchestration
 */

import { Agent, AgentResult } from '../core/agent.js';

export type TaskStatus = 'pending' | 'running' | 'completed' | 'failed' | 'cancelled';

export interface Task {
  id: string;
  type: string;
  input: any;
  status: TaskStatus;
  priority?: number;
  result?: AgentResult;
  error?: string;
  createdAt: Date;
  startedAt?: Date;
  completedAt?: Date;
  retryCount?: number;
}

export interface WorkerAgent {
  id: string;
  agent: Agent;
  capabilities: string[];
  currentTask?: Task;
  healthy: boolean;
  lastHealthCheck?: Date;
}

export interface SupervisorConfig {
  maxConcurrentTasks?: number;
  taskTimeoutMs?: number;
  maxRetries?: number;
}

export class SupervisorAgent {
  private workers: Map<string, WorkerAgent> = new Map();
  private taskQueue: Task[] = [];
  private activeTasks: Map<string, Task> = new Map();
  private completedTasks: Map<string, Task> = new Map();
  private config: Required<SupervisorConfig>;
  private isRunning: boolean = false;

  constructor(config: SupervisorConfig = {}) {
    this.config = {
      maxConcurrentTasks: config.maxConcurrentTasks ?? 5,
      taskTimeoutMs: config.taskTimeoutMs ?? 60000,
      maxRetries: config.maxRetries ?? 3
    };
  }

  registerWorker(id: string, agent: Agent, capabilities: string[]): void {
    this.workers.set(id, {
      id,
      agent,
      capabilities,
      healthy: true,
      lastHealthCheck: new Date()
    });
  }

  async submitTask(taskInput: { id: string; type: string; input: any; priority?: number }): Promise<string> {
    const task: Task = {
      ...taskInput,
      status: 'pending',
      priority: taskInput.priority ?? 0,
      createdAt: new Date(),
      retryCount: 0
    };

    const insertIndex = this.taskQueue.findIndex(t => (t.priority ?? 0) > (task.priority ?? 0));
    if (insertIndex === -1) {
      this.taskQueue.push(task);
    } else {
      this.taskQueue.splice(insertIndex, 0, task);
    }

    return task.id;
  }

  async orchestrate(): Promise<void> {
    if (this.isRunning) return;
    this.isRunning = true;

    try {
      while (this.taskQueue.length > 0 || this.activeTasks.size > 0) {
        while (
          this.taskQueue.length > 0 &&
          this.activeTasks.size < this.config.maxConcurrentTasks
        ) {
          const task = this.taskQueue[0];
          const worker = this.selectWorker(task);

          if (!worker) break;

          this.taskQueue.shift();
          await this.startTask(task, worker);
        }

        if (this.activeTasks.size > 0) {
          await new Promise(resolve => setTimeout(resolve, 100));
        }
      }
    } finally {
      this.isRunning = false;
    }
  }

  private async startTask(task: Task, worker: WorkerAgent): Promise<void> {
    task.status = 'running';
    task.startedAt = new Date();
    worker.currentTask = task;
    this.activeTasks.set(task.id, task);

    try {
      const result = await Promise.race([
        worker.agent.run(JSON.stringify(task.input)),
        new Promise<never>((_, reject) =>
          setTimeout(() => reject(new Error('Task timeout')), this.config.taskTimeoutMs)
        )
      ]);

      task.status = 'completed';
      task.result = result;
      task.completedAt = new Date();
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : String(error);
      if ((task.retryCount ?? 0) < this.config.maxRetries) {
        task.retryCount = (task.retryCount ?? 0) + 1;
        task.status = 'pending';
        this.taskQueue.unshift(task);
      } else {
        task.status = 'failed';
        task.error = errorMessage;
        task.completedAt = new Date();
      }
    } finally {
      worker.currentTask = undefined;
      this.activeTasks.delete(task.id);
      this.completedTasks.set(task.id, task);
    }
  }

  private selectWorker(task: Task): WorkerAgent | undefined {
    for (const [, worker] of this.workers) {
      if (worker.healthy && !worker.currentTask && worker.capabilities.includes(task.type)) {
        return worker;
      }
    }
    return undefined;
  }

  getTaskStatus(taskId: string): Task | undefined {
    return this.completedTasks.get(taskId) || this.activeTasks.get(taskId) || this.taskQueue.find(t => t.id === taskId);
  }
}

File: INTeract-ive-Agent/python/src/framework/orchestration/supervisor.py

import asyncio
import time
import json
from typing import TypedDict, Optional, Dict, List, Any
from src.framework.core.agent import Agent

# Type definitions
TaskStatus = str # 'pending' | 'running' | 'completed' | 'failed' | 'cancelled';

class Task(TypedDict):
    id: str
    type: str # capability needed
    input: Any
    status: TaskStatus
    priority: int
    result: Optional[Any]
    error: Optional[str]
    created_at: float
    started_at: Optional[float]
    completed_at: Optional[float]
    retry_count: int

class WorkerAgent(TypedDict):
    id: str
    agent: Agent
    capabilities: List[str]
    current_task: Optional[Task]
    healthy: bool
    last_health_check: float

class SupervisorConfig(TypedDict):
    max_concurrent_tasks: int
    task_timeout_ms: int
    max_retries: int

class SupervisorAgent:
    def __init__(self, config: Optional[SupervisorConfig] = None):
        self.config = config or {
            "max_concurrent_tasks": 5,
            "task_timeout_ms": 30000,
            "max_retries": 3
        }
        self.workers: Dict[str, WorkerAgent] = {}
        self.task_queue: List[Task] = []
        self.active_tasks: Dict[str, Task] = {}
        self.completed_tasks: Dict[str, Task] = {}
        self.is_running = False

    def register_worker(self, id: str, agent: Agent, capabilities: List[str]):
        self.workers[id] = {
            "id": id,
            "agent": agent,
            "capabilities": capabilities,
            "current_task": None,
            "healthy": True,
            "last_health_check": time.time()
        }

    async def submit_task(self, task_input: Dict[str, Any]) -> str:
        task: Task = {
            "id": task_input["id"],
            "type": task_input["type"],
            "input": task_input["input"],
            "status": "pending",
            "priority": task_input.get("priority", 0),
            "result": None,
            "error": None,
            "created_at": time.time(),
            "started_at": None,
            "completed_at": None,
            "retry_count": 0
        }
        
        inserted = False
        for i, t in enumerate(self.task_queue):
            if t["priority"] < task["priority"]:
                self.task_queue.insert(i, task)
                inserted = True
                break
        if not inserted:
            self.task_queue.append(task)
            
        return task["id"]

    async def orchestrate(self):
        if self.is_running:
            return
        
        self.is_running = True
        try:
            while self.task_queue or self.active_tasks:
                
                while self.task_queue and len(self.active_tasks) < self.config["max_concurrent_tasks"]:
                    task = self.task_queue[0]
                    worker = self._select_worker(task)
                    
                    if not worker:
                        break
                    
                    self.task_queue.pop(0)
                    asyncio.create_task(self._start_task(task, worker))
                    
                if self.active_tasks:
                    # Yield control to let tasks run
                    await asyncio.sleep(0.1)
                else:
                    break

        finally:
            self.is_running = False

    def _select_worker(self, task: Task) -> Optional[WorkerAgent]:
        for worker in self.workers.values():
            if (worker["healthy"] and 
                worker["current_task"] is None and 
                task["type"] in worker["capabilities"]):
                return worker
        return None

    async def _start_task(self, task: Task, worker: WorkerAgent):
        task["status"] = "running"
        task["started_at"] = time.time()
        worker["current_task"] = task
        self.active_tasks[task["id"]] = task

        try:
            # Prepare prompt for subagent
            if isinstance(task["input"], dict):
                prompt = json.dumps(task["input"])
            else:
                prompt = str(task["input"])

            result = await asyncio.wait_for(
                worker["agent"].run(prompt),
                timeout=self.config["task_timeout_ms"] / 1000.0
            )
            task["status"] = "completed"
            task["result"] = result["response"] # Use the text response
            task["completed_at"] = time.time()
        except Exception as e:
            error_msg = str(e)
            if task["retry_count"] < self.config["max_retries"]:
                task["retry_count"] += 1
                task["status"] = "pending"
                self.task_queue.insert(0, task)
            else:
                task["status"] = "failed"
                task["error"] = error_msg
                task["completed_at"] = time.time()
        finally:
            worker["current_task"] = None
            if task["status"] in ["completed", "failed"]:
                self.completed_tasks[task["id"]] = task
            if task["id"] in self.active_tasks:
                del self.active_tasks[task["id"]]
    
    def get_task_status(self, task_id: str) -> Optional[Task]:
        return (self.completed_tasks.get(task_id) or 
                self.active_tasks.get(task_id) or 
                next((t for t in self.task_queue if t["id"] == task_id), None))
